x-pack/plugins/security_solution/public/detections/containers/detection_engine/rules/rules_table/rules_table_context.tsx

```diff
@@ -0,0 +1,292 @@
+  /**
+   * Indicates which rule action (enable, disable, delete, etc.) is currently in progress
+   */
+  loadingRulesAction: LoadingRuleAction;
+  /**
+   * Currently selected page and number of rows per page
+   */
+  pagination: PaginationOptions;
+  /**
+   * IDs of rules selected by a user
+   */
+  selectedRuleIds: string[];
+  /**
+   * Currently selected table sorting
+   */
+  sortingOptions: SortingOptions;
+}
+
+const initialFilterOptions: FilterOptions = {
+  filter: '',
+  tags: [],
+  showCustomRules: false,
+  showElasticRules: false,
+};
+
+const initialSortingOptions: SortingOptions = {
+  field: 'enabled',
+  order: 'desc',
+};
+
+export type LoadingRuleAction =
+  | 'delete'
+  | 'disable'
+  | 'duplicate'
+  | 'enable'
+  | 'export'
+  | 'load'
+  | 'edit'
+  | null;
+
+interface LoadingRules {
+  ids: string[];
+  action: LoadingRuleAction;
+}
+
+export interface RulesTableActions {
+  reFetchRules: ReturnType<typeof useAllRules>['refetch'];
+  setFilterOptions: React.Dispatch<React.SetStateAction<FilterOptions>>;
+  setIsAllSelected: React.Dispatch<React.SetStateAction<boolean>>;
+  setIsRefreshOn: React.Dispatch<React.SetStateAction<boolean>>;
+  setLoadingRules: React.Dispatch<React.SetStateAction<LoadingRules>>;
+  setPage: React.Dispatch<React.SetStateAction<number>>;
+  setPerPage: React.Dispatch<React.SetStateAction<number>>;
+  setSelectedRuleIds: React.Dispatch<React.SetStateAction<string[]>>;
+  setSortingOptions: React.Dispatch<React.SetStateAction<SortingOptions>>;
+  updateRules: (newRules: Rule[]) => void;
+}
+
+export interface RulesTableContextType {
+  state: RulesTableState;
+  actions: RulesTableActions;
+}
+
+const RulesTableContext = createContext<RulesTableContextType | null>(null);
+
+interface RulesTableContextProviderProps {
+  children: React.ReactNode;
+  totalRules: number | null;
+  refetchPrePackagedRulesStatus: () => Promise<void>;
+}
+
+export const RulesTableContextProvider = ({
+  children,
+  totalRules,
+  refetchPrePackagedRulesStatus,
+}: RulesTableContextProviderProps) => {
+  const [autoRefreshSettings] = useUiSetting$<{
+    on: boolean;
+    value: number;
+    idleTimeout: number;
+  }>(DEFAULT_RULES_TABLE_REFRESH_SETTING);
+
+  const [advancedFilteringThreshold] = useUiSetting$<number>(
+    RULES_TABLE_ADVANCED_FILTERING_THRESHOLD
+  );
+
+  const hasTotalRules = totalRules != null;
+  const isInMemorySorting = hasTotalRules ? totalRules < advancedFilteringThreshold : false;
+
+  const [filterOptions, setFilterOptions] = useState<FilterOptions>(initialFilterOptions);
+  const [sortingOptions, setSortingOptions] = useState<SortingOptions>(initialSortingOptions);
+  const [isAllSelected, setIsAllSelected] = useState(false);
+  const [isRefreshOn, setIsRefreshOn] = useState(autoRefreshSettings.on);
+  const [loadingRules, setLoadingRules] = useState<LoadingRules>({ ids: [], action: null });
+  const [page, setPage] = useState(1);
+  const [perPage, setPerPage] = useState(20);
```

Comment:
nit: should we define `20` as constant, as it's a default page size?

Reply:
Done